ToolBox/KrakenDataDownloader/KrakenDataDownloader.cs

```diff
@@ -0,0 +1,148 @@
+
+        /// <summary>
+        /// Get historical data enumerable for a trading pair, type and resolution given this start and end time (in UTC).
+        /// </summary>
+        /// <param name="symbol">Symbol for the data we're looking for.</param>
+        /// <param name="resolution">Resolution of the data request</param>
+        /// <param name="startUtc">Start time of the data in UTC</param>
+        /// <param name="endUtc">End time of the data in UTC</param>
+        /// <returns>Enumerable of base data for this symbol</returns>
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            if (endUtc < startUtc)
+            {
+                throw new ArgumentException("The end date must be greater or equal than the start date.");
+            }
+
+            if (resolution != Resolution.Tick)
+            {
+                throw new NotSupportedException("Only Tick Resolution is supported.");
+            }
+
+            var startUnixTime = Convert.ToInt64(Time.DateTimeToUnixTimeStamp(startUtc) * 1000000000); // Multiply by 10^9 per Kraken API
+            var endUnixTime = Convert.ToInt64(Time.DateTimeToUnixTimeStamp(endUtc) * 1000000000);
+            var url = string.Format(UrlPrototype, symbol.Value, startUnixTime);
+            List<List<string>> data;
+
+            using (var client = new WebClient())
+            {
+                var rateGate = new RateGate(10, TimeSpan.FromMinutes(1)); // 10 calls per minute for Kraken API
+
+                rateGate.WaitToProceed();
+                var response = client.DownloadString(url);
+                dynamic result = JsonConvert.DeserializeObject<dynamic>(response);
+                if (result.error.Count != 0)
+                {
+                    throw new Exception("Error in Kraken API: " + result.error[0]);
+                }
+
+                data = result.result[symbol.Value].ToObject<List<List<string>>>();
+
+                foreach (var i in data)
+                {
+                    var time = Time.UnixTimeStampToDateTime(Convert.ToDouble(i[2].Split('.')[0]));
+                    if (time > endUtc)
+                    {
+                        break;
+                    }
+
+                    var value = Decimal.Parse(i[0]);
+                    var volume = Decimal.Parse(i[1]);
+
+                    yield return new Tick
+                    {
+                        Value = value,
+                        Time = time,
+                        DataType = MarketDataType.Tick,
+                        Symbol = symbol,
+                        TickType = TickType.Trade,
+                        Quantity = volume,
+                        Exchange = "kraken"
+                    };
+                }
+
+                var last = Convert.ToInt64(result.result.last);
+                while (last < endUnixTime)
```

Comment:
Can you explain why there are 2 yielding loops?

Reply:
The Kraken API only returns a subset of the set of all trades that has happened since the time the user specifies. As a result, we have to repeatedly call the API until it tells us that we have no more trades to pull from it. Having two yielding loops allows for the data to be processed and written to a file while we are waiting for the RateGate to allow us to request from Kraken more trades.